gpt/file.js

```diff
@@ -1,78 +1,152 @@
+import { Serializable } from "../load/serializable.js";
+
+export type Json =
+  | string
+  | number
+  | boolean
+  | null
+  | { [key: string]: Json }
+  | Json[];
+
+export type JsonObject = { [key: string]: Json };
+
+/**
+ * Represents a JSON object in the LangChain framework. Provides methods
+ * to get keys and values from the JSON object.
+ */
+export class JsonSpec extends Serializable {
+  lc_namespace = ["langchain", "tools", "json"];
+
+  obj: JsonObject;
+
+  maxValueLength = 4000;
+
+  constructor(obj: JsonObject, max_value_length = 4000) {
+    super(...arguments);
+    this.obj = obj;
+    this.maxValueLength = max_value_length;
   }
 
-  if (!question) {
-    throw new Error('Missing argument: question');
+  /**
+   * Retrieves all keys at a given path in the JSON object.
+   * @param input The path to the keys in the JSON object, provided as a string in JSON pointer syntax.
+   * @returns A string containing all keys at the given path, separated by commas.
+   */
+  public getKeys(input: string): string {
+    const pointer = jsonpointer.compile(input);
+    const res = pointer.get(this.obj) as Json;
+    if (typeof res === "object" && !Array.isArray(res) && res !== null) {
+      return Object.keys(res)
+        .map((i) => i.replaceAll("~", "~0").replaceAll("/", "~1"))
+        .join(", ");
+    }
+
+    throw new Error(
+      `Value at ${input} is not a dictionary, get the value directly instead.`
+    );
   }
 
-  chatHistory.add({
-    role: 'user',
-    content: `Primary Objective/User Feedback: ${question}`,
-  });
+  /**
+   * Retrieves the value at a given path in the JSON object.
+   * @param input The path to the value in the JSON object, provided as a string in JSON pointer syntax.
+   * @returns The value at the given path in the JSON object, as a string. If the value is a large dictionary or exceeds the maximum length, a message is returned instead.
+   */
+  public getValue(input: string): string {
+    const pointer = jsonpointer.compile(input);
+    const res = pointer.get(this.obj) as Json;
 
-  let chunkMaps = {};
-  for (let filePath of filePaths) {
-    console.log('Processing file:', filePath);
-    filePath = path.resolve(process.cwd(), filePath);
-    try {
-      const result = await processFile(filePath);
-      if (!chunkMaps[filePath]) chunkMaps[filePath] = [];
-      chunkMaps[filePath] = chunkMaps[filePath].concat(result.chunks);
-      //getLastItemInArray
-      // const lastItem = result.chunks[result.chunks.length - 1];
-      // console.log(lastItem.text);
-    } catch (err) {
-      console.error('Error processing file:', filePath);
-      throw err;
+    if (res === null || res === undefined) {
+      throw new Error(`Value at ${input} is null or undefined.`);
     }
+
+    const str = typeof res === "object" ? JSON.stringify(res) : res.toString();
+    if (
+      typeof res === "object" &&
+      !Array.isArray(res) &&
+      str.length > this.maxValueLength
```

Comment:
The 'JsonListKeysTool' and 'JsonGetValueTool' classes are very similar and share a lot of code. Consider creating a base class that contains the shared code and have these classes extend the base class.